### ISSUE TITLE
ORCA Panics when using the CREATE FUNCTION documentation example for "Use polymorphic types to return an ENUM array"

### ISSUE THREAD

**Issue by edespino** [greenplum-db/gpdb#12171]
**Greenplum Version**
```
postgres --gp-version
postgres (Greenplum Database) 6.16.2 build commit:950f103fe23180ed0148552ef08cd62cbbf5a681
```

**OS Version**
```
  CentOS Linux release 7.9.2009 (Core)
  Linux mdw 3.10.0-1160.6.1.el7.x86_64 #1 SMP Tue Nov 17 13:59:11 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Problem description**

From the Greenplum 6.16 documentation, using the [CREATE FUNCTION](https://greenplum.docs.pivotal.io/6-16/ref_guide/sql_commands/CREATE_FUNCTION.html) _Use polymorphic types to return an ENUM array_ example generates a server panic when the function is used in a `SELECT` statement. The problem is reliably reproducible.

Please note: **_This behavior is NOT exhibited with the planner._**

**Reproduction steps**

```
CREATE TYPE rainbow AS ENUM('red','orange','yellow','green','blue','indigo','violet');
CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement ) 
    RETURNS TABLE (ae anyenum, aa anyarray) AS $$
    SELECT $1, array[$2, $3] 
$$ LANGUAGE SQL STABLE;

SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);
```

**Here is the master server stack**

```
2021-06-16 02:15:56.065475 EDT,"gpadmin","dev",p16957,th-1744660352,"[local]",,2021-06-16 02:15:21 EDT,0,con10,cmd3,seg-1,,,,sx1,"LOG","00000","statement: CREATE TYPE rainbow AS ENUM('red','orange','yellow','green','blue','indigo','violet');",,,,,,"CREATE TYPE rainbow AS ENUM('red','orange','yellow','green','blue','indigo','violet');",0,,"postgres.c",1639,
2021-06-16 02:15:56.084572 EDT,"gpadmin","dev",p16957,th-1744660352,"[local]",,2021-06-16 02:15:21 EDT,0,con10,cmd4,seg-1,,,,sx1,"LOG","00000","statement: CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
    RETURNS TABLE (ae anyenum, aa anyarray) AS $$
    SELECT $1, array[$2, $3]
$$ LANGUAGE SQL STABLE;",,,,,,"CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
    RETURNS TABLE (ae anyenum, aa anyarray) AS $$
    SELECT $1, array[$2, $3]
$$ LANGUAGE SQL STABLE;",0,,"postgres.c",1639,
2021-06-16 02:15:56.094923 EDT,"gpadmin","dev",p16957,th-1744660352,"[local]",,2021-06-16 02:15:21 EDT,0,con10,cmd5,seg-1,,,,sx1,"LOG","00000","statement: SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);",,,,,,"SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);",0,,"postgres.c",1639,
2021-06-16 02:15:56.298684 EDT,,,p16957,th0,,,2021-06-16 02:15:21 EDT,0,con10,cmd5,seg-1,,,,,"PANIC","XX000","Unexpected internal error: Master process received signal SIGSEGV",,,,,,,0,,,,"1    0x7f099562e630 libpthread.so.0 <symbol not found> + 0x9562e630
2    0xfcec66 postgres _ZN5gpdxl16CTranslatorUtils23ConvertToCDXLLogicalTVFEPN4gpos11CMemoryPoolEPN5gpopt11CMDAccessorEPNS_12CIdGeneratorEPK13RangeTblEntry (CRefCount.h:103)
3    0xfef497 postgres _ZN5gpdxl21CTranslatorQueryToDXL17TranslateTVFToDXLEPK13RangeTblEntryjj (CTranslatorQueryToDXL.cpp:3577)
4    0xfe88f6 postgres _ZN5gpdxl21CTranslatorQueryToDXL22TranslateFromExprToDXLEP8FromExpr (CTranslatorQueryToDXL.cpp:2922)
5    0xfe9d7a postgres _ZN5gpdxl21CTranslatorQueryToDXL21TranslateGroupingSetsEP8FromExprP4ListS4_bPN4gpos8CHashMapIijXadL_ZNS5_9HashValueIiEEjPKT_EEXadL_ZNS5_6EqualsIiEEbSA_SA_EEXadL_ZNS5_13CleanupDeleteIiEEvPS8_EEXadL_ZNSC_IjEEvSD_EEEESF_ (CTranslatorQueryToDXL.cpp:613)
6    0xfea1bd postgres _ZN5gpdxl21CTranslatorQueryToDXL25TranslateSelectQueryToDXLEv (CTranslatorQueryToDXL.cpp:557)
7    0xfeb946 postgres _ZN5gpdxl21CTranslatorQueryToDXL19TranslateQueryToDXLEv (CTranslatorQueryToDXL.cpp:659)
8    0x10004f4 postgres _ZN9COptTasks12OptimizeTaskEPv (discriminator 3)
9    0xcf12d6 postgres _ZN4gpos5CTask7ExecuteEv (CTask.cpp:130)
10   0xcf2268 postgres _ZN4gpos7CWorker7ExecuteEPNS_5CTaskE (CWorker.cpp:82)
11   0xcf079d postgres _ZN4gpos14CAutoTaskProxy7ExecuteEPNS_5CTaskE (CAutoTaskProxy.cpp:301)
12   0xcf2e61 postgres gpos_exec (_api.cpp:246)
13   0xffe6a1 postgres _ZN9COptTasks7ExecuteEPFPvS0_ES0_ (COptTasks.cpp:261)
14   0xfff8ce postgres _ZN9COptTasks18GPOPTOptimizedPlanEP5QueryP11SOptContext (COptTasks.cpp:775)
15   0x1001b91 postgres GPOPTOptimizedPlan (CGPOptimizer.cpp:60)
16   0x9d3243 postgres optimize_query (orca.c:46)
17   0x9c48cd postgres standard_planner (planner.c:240)
18   0x9c4c25 postgres planner (planner.c:200)
19   0xa8ac01 postgres <symbol not found> (postgres.c:961)
20   0xa8d59f postgres PostgresMain (postgres.c:5242)
"
2021-06-16 02:15:56.298876 EDT,,,p15522,th-1744660352,,,,0,,,seg-1,,,,,"LOG","00000","server process (PID 16957) was terminated by signal 11: Segmentation fault","Failed process was running: SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);",,,,,,0,,"postmaster.c",3970,
```

**Comment by chrishajas** [comment]
This is identical to https://github.com/greenplum-db/gpdb/issues/11880, it will be fixed with https://github.com/greenplum-db/gpdb/pull/12051. Note that we haven't been able to repro this on local machines, and it does appear to be OS/system dependent.

**Comment by chrishajas** [comment]
Merged fix to 6X (and master) in 2f45387.
